DOC-1895: Added RN for v21.2.8 (#13442)
cockroachdb/docs: _includes/releases/v21.2/v21.2.8.md

## v21.2.8

Release Date: April 4, 2022

{% include releases/release-downloads-docker-image.md release=include.release %}

<h3 id="v21-2-8-security-updates">Security updates</h3>

- Users can enable HSTS headers to be set on all HTTP requests, which force browsers to upgrade to HTTPS without a redirect. This is controlled by setting the `server.hsts.enabled` [cluster setting](../v21.2/cluster-settings.html), which is `false` by default, to `true`. [#77845][#77845]

> - Add a new flag `--external-io-enable-non-admin-implicit-access` that can remove the admin-only restriction on interacting with arbitrary network endpoints and using 'implicit' authentication in operations such as [`BACKUP`](../v21.2/backup.html), [`IMPORT`](../v21.2/import.html), or [`EXPORT`](../v21.2/export.html). [#78599][#78599]

Reviewer: ```suggestion
- Added a new flag `--external-io-enable-non-admin-implicit-access` that can remove the admin-only restriction on interacting with arbitrary network endpoints and using implicit authentication in operations such as [`BACKUP`](../v21.2/backup.html), [`IMPORT`](../v21.2/import.html), or [`EXPORT`](../v21.2/export.html). [#78599][#78599]
```

<h3 id="v21-2-8-enterprise-edition-changes">Enterprise edition changes</h3>

- Changefeeds running on tables with a low value for the `gc.ttlseconds` [cluster setting](../v21.2/cluster-settings.html) will function more reliably due to protected timestamps being maintained for the changefeed targets at the resolved timestamp of the changefeed. The frequency at which the protected timestamp is updated to the resolved timestamp can be configured through the `changefeed.protect_timestamp_interval` cluster setting. If the changefeed lags too far behind such that storage of old data becomes an issue, cancelling the changefeed will release the protected timestamps and allow garbage collection to resume. If `protect_data_from_gc_on_pause` is unset, pausing the changefeed will release the existing protected timestamp record. [#77589][#77589]
- Added the "changefeed.backfill_pending_ranges" prometheus metric to track ongoing backfill progress of a changefeed. [#77383][#77383]

> - Changefeeds record message size histograms. [#77932][#77932]

Reviewer: ```suggestion
- Changefeeds now record message size histograms. [#77932][#77932]
```

> - Limit the number of concurent catchup scan requests issued by rangefeed client. [#77932][#77932]

Reviewer: ```suggestion
- The number of concurrent catchup scan requests issued by a rangefeed client is now limited. [#77932][#77932]
```

> - Remove expensive, unnecessary, and never used `schedules.round.schedules-ready-to-run` and `schedules.round.num-jobs-running` metrics from job schedulers. [#78583][#78583]

Reviewer: ```suggestion
- Removed the expensive and unnecessary `schedules.round.schedules-ready-to-run` and `schedules.round.num-jobs-running` metrics from job schedulers. [#78583][#78583]
```

<h3 id="v21-2-8-sql-language-changes">SQL language changes</h3>

> - Added a `sql.auth.resolve_membership_single_scan.enabled` [cluster setting](../v21.2/cluster-settings.html), which changes the query for an internal role membership cache. Previously the code would recursively look up each role in the membership hierarchy, leading to multiple queries. With the setting on, it uses a single query. The setting is `false` by default. [#77631][#77631]

Reviewer: ```suggestion
- Added a `sql.auth.resolve_membership_single_scan.enabled` [cluster setting](../v21.2/cluster-settings.html), which changes the query for an internal role membership cache. Previously the code would recursively look up each role in the membership hierarchy, leading to multiple queries. With the setting on, it uses a single query. This setting is `false` by default. [#77631][#77631]
```

> - When the user runs [`SHOW BACKUP`](../v21.2/show-backup.html) on an encrypted incremental backup, they must set the `encyrption_info_dir` directory to the full backup directory in order for `SHOW BACKUP` to work. [#78141][#78141]

Reviewer: ```suggestion
- When users run [`SHOW BACKUP`](../v21.2/show-backup.html) on an encrypted incremental backup, they must now set the `encyrption_info_dir` directory to the full backup directory in order for `SHOW BACKUP` to work. [#78141][#78141]
```

> - The stats compaction scheduled job no longer causes intent buildup. [#78583][#78583]

Reviewer: ```suggestion
- The [stats compaction](../v21.2/cost-based-optimizer#table-statistics) scheduled job no longer causes intent buildup. [#78583][#78583]
```

> - Initial implementation of a scheduled logger used to capture index usage statistics to the telemetry logging channel. [#78522][#78522]

Reviewer: ```suggestion
- Implemented a scheduled logger used to capture index usage statistics to the telemetry logging channel. [#78522][#78522]
```

<h3 id="v21-2-8-operational-changes">Operational changes</h3>

- The setting `kv.bulk_io_write.min_capacity_remaining_fraction` can now be set to cause bulk ingest operations like [`IMPORT`](../v21.2/import.html), [`RESTORE`](../v21.2/restore.html), or [`CREATE INDEX`](../v21.2/create-index.html) to fail rather than write to a node that is running out of disk space. [#78575][#78575]
- Improved jobs system resilience to scheduled jobs that may lock up jobs/scheduled job table for long periods of time. Each schedule now has a limited amount of time to complete its execution. The timeout is controlled via the `jobs.scheduler.schedule_execution.timeout` [cluster setting](../v21.2/cluster-settings.html). [#77620][#77620]

<h3 id="v21-2-8-command-line-changes">Command-line changes</h3>

> - - The `cockroach debug tsdump` command now allows viewing time-series data even in case of node failures by being re-run with the import filename set to `-`. [#77976][#77976]

Reviewer: ```suggestion
- The `cockroach debug tsdump` command now allows viewing timeseries data even in cases of node failure by allowing users to rerun the command with the import filename set to `"-"`. [#77976][#77976]
```

> - Fixes a bug where running [`cockroach demo`](../v21.2/cockroach-demo.html) with the `--global` flag would not simulate latencies correctly when combined with the `--insecure` flag. [#78170][#78170]

Reviewer: ```suggestion
- Fixed a bug where running [`cockroach demo`](../v21.2/cockroach-demo.html) with the `--global` flag would not simulate latencies correctly when combined with the `--insecure` flag. [#78170][#78170]
```

<h3 id="v21-2-8-db-console-changes">DB Console changes</h3>

> - Add **Hot Ranges** page and link to it on the sidebar. [#77594][#77594]

Reviewer: ```suggestion
- Added a **Hot Ranges** page and linked to it in the sidebar. [#77594][#77594]
```

> - The `_status/nodes` endpoint is avaible to all users with the `VIEWACTIVITY` role option, not just Admins. In the **DB Console**, the **Nodes Overview**, and the **Node Reports** pages will now display unredacted information containing node hostnames and IP addresses for all users with the `VIEWACTIVITY` role option. Previously this was also gated for Admins only. [#78275][#78275]

Reviewer: ```suggestion
- The `_status/nodes` endpoint is now available to all users with the `VIEWACTIVITY` role option, not just `admin` users. In the DB Console, the **Nodes Overview** and **Node Reports** pages will now display unredacted information containing node hostnames and IP addresses for all users with the `VIEWACTIVITY` role option. [#78275][#78275]
```

> - Fixes a bug where a node in the `UNAVAILABLE` state will not have latency defined and cause the network page to crash. [#78627][#78627]

Reviewer: ```suggestion
- Fixed a bug where a node in the `UNAVAILABLE` state would not have latency defined and caused the network page to crash. [#78627][#78627]
```

<h3 id="v21-2-8-bug-fixes">Bug fixes</h3>

- Fixed a bug that caused errors when attempting to create table statistics (with [`CREATE STATISTICS`](../v21.2/create-statistics.html) or [`ANALYZE`](../v21.2/explain-analyze.html)) for a table containing an index that indexed only virtual [`computed columns`](../v21.2/computed-columns.html). This bug has been present since version 21.1.0. [#77565][#77565]
- Fixed a bug when adding a [`hash-sharded index`](../v21.2/hash-sharded-indexes.html) to a table watched by a changefeed. [#77739][#77739]
- Fixed successive schema change backfills from skipping spans that were checkpointed by an initial backfill that was restarted. [#77829][#77829]

> - Attempting to run concurrent profiles works up to a concurrency limit of two. This will remove the occurrence of "profile id not found" errors while running upto two profiles concurrently. When a profile is not found, the error message has been updated to suggest remediation steps in order to unblock the user. [#77977][#77977]

Reviewer: ```suggestion
- Attempting to run concurrent profiles now works up to a concurrency limit of two. This will remove the occurrence of `profile id not found` errors while running up to two profiles concurrently. When a profile is not found, the error message has been updated to suggest troubleshooting steps. [#77977][#77977]
```

- Fixed an optimizer bug that prevented expressions of the form `(NULL::STRING[] <@ ARRAY['x'])` from being folded to `NULL`. This bug was introduced in v21.2.0. [#78039][#78039]

> - Added a limit of 7 concurrent asynchronous consistency checks per store, with an upper timeout of 1 hour. This prevents abandoned consistency checks from building up in some circumstances, which could lead to increasing disk usage as they held onto Pebble snapshots. [#77611][#77611]

Reviewer: ```suggestion
- Added a limit of seven concurrent asynchronous consistency checks per store, with an upper timeout of one hour. This prevents abandoned consistency checks from building up in some circumstances, which could lead to increasing disk usage as they held onto Pebble snapshots. [#77611][#77611]
```

- Fixed a bug where the [**Statement Details**](../v21.2/ui-statements-page.html#statement-details-page) page fails to load query plan even after when the plan has been sampled. [#78105][#78105]
- Fixed a memory leak in the Pebble block cache. [#78257][#78257]

> - Fixed a bug that caused internal errors when COALESCE and IF expressions had inner expressions with different types that could not be cast to a common type. [#78342][#78342]

Reviewer: ```suggestion
- Fixed a bug that caused internal errors when `COALESCE` and `IF` expressions had inner expressions with different types that could not be cast to a common type. [#78342][#78342]
```

> - CockroachDB might now fetch less rows when performing lookup and index joins on the queries with [`LIMIT`](../v21.2/limit-offset.html#limit) clause. [#78474][#78474]

Reviewer: ```suggestion
- CockroachDB might now fetch fewer rows when performing lookup and index joins on queries with the [`LIMIT`](../v21.2/limit-offset.html#limit) clause. [#78474][#78474]
```

- Added a new flag `--external-io-enable-non-admin-implicit-access` that can remove the admin-only restriction on interacting with arbitrary network endpoints and using implicit authentication in operations such as [`BACKUP`](../v21.2/backup.html), [`IMPORT`](../v21.2/import.html), or [`EXPORT`](../v21.2/export.html). [#78599][#78599]

<h3 id="v21-2-8-enterprise-edition-changes">Enterprise edition changes</h3>

- Changefeeds running on tables with a low value for the `gc.ttlseconds` [cluster setting](../v21.2/cluster-settings.html) will function more reliably due to protected timestamps being maintained for the changefeed targets at the resolved timestamp of the changefeed. The frequency at which the protected timestamp is updated to the resolved timestamp can be configured through the `changefeed.protect_timestamp_interval` cluster setting. If the changefeed lags too far behind such that storage of old data becomes an issue, cancelling the changefeed will release the protected timestamps and allow garbage collection to resume. If `protect_data_from_gc_on_pause` is unset, pausing the changefeed will release the existing protected timestamp record. [#77589][#77589]

> - Added the "changefeed.backfill_pending_ranges" prometheus metric to track ongoing backfill progress of a changefeed. [#77383][#77383]

Reviewer: ```suggestion
- Added the `changefeed.backfill_pending_ranges` prometheus metric to track the ongoing backfill progress of a changefeed. [#77383][#77383]
```

- When users run [`SHOW BACKUP`](../v21.2/show-backup.html) on an encrypted incremental backup, they must now set the `encyrption_info_dir` directory to the full backup directory in order for `SHOW BACKUP` to work. [#78141][#78141]
- The [stats compaction](../v21.2/cost-based-optimizer.html#table-statistics) scheduled job no longer causes intent buildup. [#78583][#78583]
- Implemented a scheduled logger used to capture index usage statistics to the telemetry logging channel. [#78522][#78522]

<h3 id="v21-2-8-operational-changes">Operational changes</h3>

- The setting `kv.bulk_io_write.min_capacity_remaining_fraction` can now be set to cause bulk ingest operations like [`IMPORT`](../v21.2/import.html), [`RESTORE`](../v21.2/restore.html), or [`CREATE INDEX`](../v21.2/create-index.html) to fail rather than write to a node that is running out of disk space. [#78575][#78575]

> - Improved jobs system resilience to scheduled jobs that may lock up jobs/scheduled job table for long periods of time. Each schedule now has a limited amount of time to complete its execution. The timeout is controlled via the `jobs.scheduler.schedule_execution.timeout` [cluster setting](../v21.2/cluster-settings.html). [#77620][#77620]

Reviewer: ```suggestion
- Improved jobs system resilience to scheduled jobs that may lock up the scheduled jobs table for long periods of time. Each schedule now has a limited amount of time to complete its execution. The timeout is controlled via the `jobs.scheduler.schedule_execution.timeout` [cluster setting](../v21.2/cluster-settings.html). [#77620][#77620]
```

<h3 id="v21-2-8-command-line-changes">Command-line changes</h3>

- The `cockroach debug tsdump` command now allows viewing timeseries data even in cases of node failure by allowing users to rerun the command with the import filename set to `"-"`. [#77976][#77976]
- Fixed a bug where running [`cockroach demo`](../v21.2/cockroach-demo.html) with the `--global` flag would not simulate latencies correctly when combined with the `--insecure` flag. [#78170][#78170]

<h3 id="v21-2-8-db-console-changes">DB Console changes</h3>

- Added a **Hot Ranges** page and linked to it in the sidebar. [#77594][#77594]
- The `_status/nodes` endpoint is now available to all users with the `VIEWACTIVITY` role option, not just `admin` users. In the DB Console, the **Nodes Overview** and **Node Reports** pages will now display unredacted information containing node hostnames and IP addresses for all users with the `VIEWACTIVITY` role option. [#78275][#78275]
- Fixed a bug where a node in the `UNAVAILABLE` state would not have latency defined and caused the network page to crash. [#78627][#78627]

<h3 id="v21-2-8-bug-fixes">Bug fixes</h3>

> - Fixed a bug that caused errors when attempting to create table statistics (with [`CREATE STATISTICS`](../v21.2/create-statistics.html) or [`ANALYZE`](../v21.2/explain-analyze.html)) for a table containing an index that indexed only virtual [`computed columns`](../v21.2/computed-columns.html). This bug has been present since version 21.1.0. [#77565][#77565]

Reviewer: ```suggestion
- Fixed a bug that caused errors when attempting to create table statistics (with [`CREATE STATISTICS`](../v21.2/create-statistics.html) or [`ANALYZE`](../v21.2/explain-analyze.html)) for a table containing an index that indexed only virtual [`computed columns`](../v21.2/computed-columns.html). This bug has been present since v21.1.0. [#77565][#77565]
```

- Fixed a bug when adding a [`hash-sharded index`](../v21.2/hash-sharded-indexes.html) to a table watched by a changefeed. [#77739][#77739]
- Fixed successive schema change backfills from skipping spans that were checkpointed by an initial backfill that was restarted. [#77829][#77829]
- Attempting to run concurrent profiles now works up to a concurrency limit of two. This will remove the occurrence of `profile id not found` errors while running up to two profiles concurrently. When a profile is not found, the error message has been updated to suggest troubleshooting steps. [#77977][#77977]
- Fixed an optimizer bug that prevented expressions of the form `(NULL::STRING[] <@ ARRAY['x'])` from being folded to `NULL`. This bug was introduced in v21.2.0. [#78039][#78039]

> - Added a limit of seven concurrent asynchronous consistency checks per store, with an upper timeout of one hour. This prevents abandoned consistency checks from building up in some circumstances, which could lead to increasing disk usage as they held onto Pebble snapshots. [#77611][#77611]

Reviewer: ```suggestion
- Added a limit of seven concurrent asynchronous consistency checks per store, with an upper timeout of one hour. This prevents abandoned consistency checks from building up in some circumstances, which could lead to increasing disk usage as they held onto [Pebble](../v21.2/architecture/storage-layer.html#pebble) snapshots. [#77611][#77611]
```

- Fixed a bug where the [**Statement Details**](../v21.2/ui-statements-page.html#statement-details-page) page fails to load query plan even after when the plan has been sampled. [#78105][#78105]
- Fixed a memory leak in the Pebble block cache. [#78257][#78257]
- Fixed a bug that caused internal errors when `COALESCE` and `IF` expressions had inner expressions with different types that could not be cast to a common type. [#78342][#78342]
- CockroachDB might now fetch fewer rows when performing lookup and index joins on queries with the [`LIMIT`](../v21.2/limit-offset.html#limit) clause. [#78474][#78474]

> - A zone config change event now includes the correct details of what was changed instead of incorrectly displaying "undefined". [#78634][#78634]

Reviewer: ```suggestion
- A zone config change event now includes the correct details of what was changed instead of incorrectly displaying `undefined`. [#78634][#78634]
```

- Fixed a bug that prevented a table created on a 22.1 node to be queried on a 21.2 node in a mixed-version cluster. [#78657][#78657]

> - Fixed a bug that caused errors when trying to evaluate queries with `NULL` values annotated as a tuple type, such as `NULL:::RECORD`. This bug was present since version 19.1. [#78635][#78635]

Reviewer: ```suggestion
- Fixed a bug that caused errors when trying to evaluate queries with `NULL` values annotated as a tuple type, such as `NULL:::RECORD`. This bug was present since v19.1. [#78635][#78635]
```

- Fixed a bug where CockroachDB could lose `INT2VECTOR` and `OIDVECTOR` types of some arrays. [#78630][#78630]

> - Fixed a bug that caused the optimizer to generate invalid query plans which could result in incorrect query results. The bug, which has been present since version 21.1.0, can appear if all of the following conditions are true: 1) the query contains a semi-join, such as queries in the form: `SELECT * FROM t1 WHERE EXISTS (SELECT * FROM t2 WHERE t1.a = t2.a);`, 2) the inner table has an index containing the equality column, like `t2.a` in the example query, 3) the index contains one or more columns that prefix the equality column, and 4) the prefix columns are `NOT NULL` and are constrained to a set of constant values via a `CHECK` constraint or an `IN` condition in the filter. [#78975][#78975]

Reviewer: ```suggestion
- Fixed a bug that caused the optimizer to generate invalid query plans which could result in incorrect query results. The bug, which has been present since v21.1.0, can appear if all of the following conditions are true: 1) the query contains a semi-join, such as queries in the form: `SELECT * FROM t1 WHERE EXISTS (SELECT * FROM t2 WHERE t1.a = t2.a);`, 2) the inner table has an index containing the equality column, like `t2.a` in the example query, 3) the index contains one or more columns that prefix the equality column, and 4) the prefix columns are `NOT NULL` and are constrained to a set of constant values via a `CHECK` constraint or an `IN` condition in the filter. [#78975][#78975]
```

> - Fixed a bug where [`IMPORT INTO`](../v21.2/import-into.html) could create duplicate entries violating [`UNIQUE`](../v21.2/unique.html) constraints in [`REGIONAL BY ROW` tables](../v21.2/multiregion-overview.html#regional-by-row-tables) and tables utilizing UNIQUE WITHOUT INDEX constraints. A new post-`IMPORT` validation step for those tables now fails and rolls back the `IMPORT` in such cases. [#78975][#78975]

Reviewer: ```suggestion
- Fixed a bug where [`IMPORT INTO`](../v21.2/import-into.html) could create duplicate entries violating [`UNIQUE`](../v21.2/unique.html) constraints in [`REGIONAL BY ROW` tables](../v21.2/multiregion-overview.html#regional-by-row-tables) and tables utilizing `UNIQUE WITHOUT INDEX` constraints. A new post-`IMPORT` validation step for those tables now fails and rolls back the `IMPORT` in such cases. [#78975][#78975]
```
